### PR TITLE
Upgrade CPS

### PR DESCRIPTION
**Pull request by drewnoakes**
Includes temporary suppression of some warnings due to incorrect CPS annotations. CPS PR to fix those is here: https://devdiv.visualstudio.com/DevDiv/_git/CPS/pullrequest/221897

This also upgrades the VS threading analyzer, which required code changes.

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Build/PublishItemsOutputGroupProvider.cs [hunk 1]
```diff
@@ -24,11 +24,11 @@ internal class PublishItemsOutputGroupProvider : IOutputGroupProvider
         private readonly ConfiguredProject _configuredProject;
 
         [ImportingConstructor]
-        internal PublishItemsOutputGroupProvider(IProjectAccessor projectAccessor, ConfiguredProject configuredProject)
+        internal PublishItemsOutputGroupProvider(IProjectAccessor projectAccessor, ConfiguredProject configuredProject, JoinableTaskFactory joinableTaskFactory)
         {
             _projectAccessor = projectAccessor;
             _configuredProject = configuredProject;
-            _outputGroups = new AsyncLazy<IImmutableSet<IOutputGroup>>(GetOutputGroupMetadataAsync);
+            _outputGroups = new AsyncLazy<IImmutableSet<IOutputGroup>>(GetOutputGroupMetadataAsync, joinableTaskFactory);
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Build/PublishItemsOutputGroupProvider.cs]: @davkean please review this change.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreUnconfiguredInputDataSource.cs [hunk 2]
```diff
@@ -44,7 +44,7 @@ protected override IDisposable LinkExternalInput(ITargetBlock<RestoreInfo> targe
                 _project.Services,
                 projects => projects.Select(project => project.Services.ExportProvider.GetExportedValueOrDefault<IPackageRestoreConfiguredInputDataSource>())
                                     .Where(dataSource => dataSource != null)    // Filter out those without PackageReference
-                                    .Select(DropConfiguredProjectVersions),
+                                    .Select(DropConfiguredProjectVersions!),
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreUnconfiguredInputDataSource.cs]: Why?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UI/AddItemDialogService.cs [hunk 3]
```diff
@@ -85,7 +85,7 @@ private async Task<bool> ShowDialogAsync(IProjectTree node, __VSADDITEMFLAGS fla
                 return false;
 
             if (result.Failed)
-                throw result.Exception;
+                throw result.Exception!;
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UI/AddItemDialogService.cs]: `throw null` warns?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Build/PublishItemsOutputGroupProvider.cs (hunk 1 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Build/PublishItemsOutputGroupProvider.cs]: This isn't correct and won't run. There is no importable JoinableTaskFactory, it needs to import IProjectThreadingService.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UI/AddItemDialogService.cs (hunk 3 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UI/AddItemDialogService.cs]: Yes. Throwing `null` results in a NRE.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreUnconfiguredInputDataSource.cs (hunk 2 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreUnconfiguredInputDataSource.cs]: The `Where` doesn't convert from `T?` to `T` unfortunately.

Updated in 360f28cb8 to use a new `WhereNotNull` extension method that propagates this information.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Build/PublishItemsOutputGroupProvider.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Build/PublishItemsOutputGroupProvider.cs]: Fixed.

**Comment by drewnoakes** [comment]
Addressed feedback. Will merge once builds pass.

**Pull request merged**